include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,394 @@ class Graph final {
     //!< same schema (and same order!) as outEdges
     std::vector<std::vector<edgeid>> outEdgeIds;
 
+private:
+    // base class for all node attribute storages with attribute type info
+    // independent of the attribute type, holds bookkeeping info only:
+    // - attribute name
+    // - type info of derived (real storage holding) classes
+    // - which indices are valid
+    // - number of valid indices
+    // - the associated graph (who knows, which nodes exist)
+    class NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorageBase(const Graph *graph, std::string name, std::type_index type)
+            : name{std::move(name)}, type{type}, theGraph{graph} {}
+
+        virtual ~NodeAttributeStorageBase() = default;
+
+        virtual void invalidateAttributes() = 0;
+
+        const std::string &getName() const noexcept { return name; }
+
+        std::type_index getType() const noexcept { return type; }
+
+        bool isValid(index i) const noexcept { return i < valid.size() && valid[i]; }
+
+        // Called by Graph when node is deleted.
+        void invalidate(index i) {
+            if (isValid(i)) {
+                valid[i] = false;
+                --validElements;
+            }
+        }
+
+    protected:
+        void markValid(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (i >= valid.size()) {
+                valid.resize(i + 1);
+            }
+            valid[i] = true;
+            ++validElements;
+        }
+
+        void checkIndex(index i) const {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (!isValid(i)) {
+                throw std::runtime_error("Invalid attribute value");
+            }
+        }
+
+    private:
+        std::string name;
+        std::type_index type;
+        std::vector<bool> valid; // For each node: whether attribute is set or not.
+    protected:
+        index validElements = 0;
+        const Graph *theGraph;
+    }; // class NodeAttributeStorageBase
+
+    template <typename T>
+    class NodeAttribute;
+
+    template <typename T>
+    class NodeAttributeStorage : public Graph::NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorage(const Graph *theGraph, std::string name)
+            : NodeAttributeStorageBase{theGraph, std::move(name), typeid(T)} {}
+
+        ~NodeAttributeStorage() override { invalidateAttributes(); }
+
+        void invalidateAttributes() override {
+            for (auto &att : attrSet)
+                att->invalidateAttribute();
+        }
+
+        void resize(index i) {
```

Comment:
Here as well, and also for other methods that take a node as an input parameter.